### ISSUE TITLE
Integrate quick fixes in pairing

### ISSUE THREAD

**Issue by froelich** [dedis/kyber#305]
Hello,
We are currently using the pairing suite from kyber. In order to use UnLynx with either ed25519 or bn256, we had to do some changes and we wanted to know if these changes could be integrated in kyber or if there is a better way to do some of them. Here is the list of our quick fixes:
1. The function String() in dedis/kyber/pairing/bn256/curve.go has to be changed to:
`func (c *curvePoint) String() string {`
	`c.MakeAffine()`
	`x, y := &gfP{}, &gfP{}`
	`montDecode(x, &c.x)`
	`montDecode(y, &c.y)`
	`return x.String() + y.String()`
`}`

2. The function String() in dedis/kyber/pairing/point.go has to be changed to:
`func (p *pointG1) String() string {`
	`return p.g.String()`
`}`

--> the parentheses and the "," that were used in the first function and the "bn..." used in the second made all the simulations fail with this error:
`E : (network.(*Router).Start.func1: 104) - receive server identity failed: Error while receiving ServerIdentity during negotiation Unknown Error`
as the identity contains the public key which is written with the String functions.

3. Above the function New() in dedis/kyber/pairing/bn256/suite.go we changed this:
`// Not used other than for reflect.TypeOf()`
`var aScalar mod.Int`
`var aPointG1 pointG1`
`var aPointG2 pointG2`
`var aPointGT pointGT`

to this:
`// Not used other than for reflect.TypeOf()`
`var aScalar kyber.Scalar`
`var aPointG1 kyber.Point`
`var aPointG2 kyber.Point`
`var aPointGT kyber.Point`

otherwise we got this error:
`panic: unsupported null pointer type: kyber.Scalar

goroutine 189 [running]:
github.com/dedis/fixbuf.(*decoder).value(0xc420061428, 0x14dad20, 0xc4200bd4a0, 0x194, 0x3, 0xc4200bd4a0, 0x194)
	/Users/froelich/Documents/Go/src/github.com/dedis/fixbuf/encoding.go:106 +0x9ea
github.com/dedis/fixbuf.(*decoder).value(0xc420061428, 0x144d360, 0xc42043f980, 0x97, 0x2, 0x97, 0xc42043f9e0)
	/Users/froelich/Documents/Go/src/github.com/dedis/fixbuf/encoding.go:134 +0x19a
github.com/dedis/fixbuf.(*decoder).value(0xc420061428, 0x1474f20, 0xc42013b6e0, 0x194, 0x1, 0xc42013b6e0, 0x194)
	/Users/froelich/Documents/Go/src/github.com/dedis/fixbuf/encoding.go:116 +0x5c0
github.com/dedis/fixbuf.(*decoder).value(0xc420061428, 0x144e560, 0xc42043f9e0, 0x97, 0x0, 0x0, 0xc4202b1448)
	/Users/froelich/Documents/Go/src/github.com/dedis/fixbuf/encoding.go:134 +0x19a
github.com/dedis/fixbuf.BinaryEncoding.Read(0x156ef20, 0xc420092fc0, 0x144e560, 0x156ef40, 0xc420162a80, 0xc4200614a8, 0x1, 0x1, 0x0, 0xc4202b1550)
	/Users/froelich/Documents/Go/src/github.com/dedis/fixbuf/encoding.go:79 +0xdc
github.com/dedis/fixbuf.Read(0x156ef40, 0xc420162a80, 0x156ef20, 0xc420092fc0, 0xc42013b6e0, 0x6, 0x6, 0xc4202b15d0, 0x10125f9)
	/Users/froelich/Documents/Go/src/github.com/dedis/fixbuf/encoding.go:237 +0x112
github.com/dedis/kyber/pairing/bn256.(*Suite).Read(0xc420092fc0, 0x156ef40, 0xc420162a80, 0xc42013b6e0, 0x6, 0x6, 0xc42043f9c0, 0x0)
	/Users/froelich/Documents/Go/src/github.com/dedis/kyber/pairing/bn256/suite.go:87 +0x70
github.com/dedis/kyber/proof.(*hashProver).PriRand(0xc4202ee100, 0xc42013b6e0, 0x6, 0x6, 0xa, 0xa)
	/Users/froelich/Documents/Go/src/github.com/dedis/kyber/proof/hash.go:68 +0x73
github.com/dedis/kyber/shuffle.(*PairShuffle).Prove(0xc4201fd6c0, 0xc4203f44b0, 0xa, 0xa, 0x0, 0x0, 0x1576de0, 0xc4206a2390, 0xc4200bc000, 0xa, ...)
	/Users/froelich/Documents/Go/src/github.com/dedis/kyber/shuffle/pair.go:151 +0x47d
github.com/lca1/unlynx/lib.shuffleProofCreation.func2(0x1572780, 0xc4202ee100, 0x1517dc5, 0xb)
	/Users/froelich/Documents/Go/src/github.com/lca1/unlynx/lib/proofs.go:508 +0x133
github.com/dedis/kyber/proof.HashProve(0x1e76b30, 0xc4201005c0, 0x1517dc5, 0xb, 0xc4201bc000, 0x1, 0xc4200bc000, 0xa, 0xa, 0xc420246460)
	/Users/froelich/Documents/Go/src/github.com/dedis/kyber/proof/hash.go:138 +0x71
github.com/lca1/unlynx/lib.shuffleProofCreation(0xc4200c0480, 0xa, 0x10, 0xc42041c000, 0xa, 0xa, 0xc4201241e0, 0xa, 0xa, 0xc4203f44b0, ...)
	/Users/froelich/Documents/Go/src/github.com/lca1/unlynx/lib/proofs.go:511 +0x671
github.com/lca1/unlynx/lib.ShufflingProofCreation(0xc4200c0480, 0xa, 0x10, 0xc42041c000, 0xa, 0xa, 0x0, 0x0, 0x1576de0, 0xc4206a2390, ...)
	/Users/froelich/Documents/Go/src/github.com/lca1/unlynx/lib/proofs.go:520 +0x117
github.com/lca1/unlynx/protocols.(*ShufflingProtocol).Start(0xc420248460, 0x0, 0x0)
	/Users/froelich/Documents/Go/src/github.com/lca1/unlynx/protocols/shuffling_protocol.go:172 +0xaa5
created by github.com/lca1/unlynx/services/lemal.(*ServiceLeMal).StartProtocol
	/Users/froelich/Documents/Go/src/github.com/lca1/unlynx/services/lemal/service.go:348 +0x2d8`

4. This last change should probably not be integrated and we did it to be able to do: `suites.MustFind("bn256.g1")` in order to change from one curve to the other easily.
To do that we changed dedis/kyber/suites/all.go to:

`package suites

import (
"github.com/dedis/kyber/group/edwards25519"
"github.com/dedis/kyber/pairing"
"github.com/dedis/kyber/pairing/bn256"
"github.com/dedis/kyber"
)
// ChimeraSuite is the fancy name of the suite create from the pairing curve of kyber
type ChimeraSuite struct {
	pairing.Suite
	kyber.Group
}

// SuiTe is the instantiation of the suite
var    SuiTe = ChimeraSuite{
	Suite: bn256.NewSuiteBN256(),
	Group: bn256.NewSuiteBN256().G1(),
}

func init() {
	register(edwards25519.NewBlakeSHA256Ed25519())
	register(SuiTe)
}`

**Comment by JoaoAndreSa** [comment]
Furthermore there is something else which is strange... While running some fo the UnLynx protocols with this bn256 curve I got stuck multiple times. My intuition was correct and there were some race conditions between our test function and the overlay more specifically in this MakeAffine() function in 'bn256/curve.go'. This function is used by curvePoint.String().

```golang
func (c* curvePoint) String() string {
	c.MakeAffine()
	x, y := &gfP{}, &gfP{}
	montDecode(x, &c.x)
	montDecode(y, &c.y)
	return x.String() + y.String()
}
```

Apparently this String() function is doing something with the point that changes it's state and causes this conflict. So if we try to get a string representation on our test function while running some protocols we get this race conditions. 

I fixed the problem by simply removing the reference (*) in the curvePoint.

```golang
func (c curvePoint) String() string {
	c.MakeAffine()
	x, y := &gfP{}, &gfP{}
	montDecode(x, &c.x)
	montDecode(y, &c.y)
	return x.String() + y.String()
}
```
